File: Notification.java
package com.github.jnstockley;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class Notification<E> {

	private String title;

	private String message;

	private String url;

	public NotificationType type;

	public Notification(List<E> data) {
		if (data.get(0).getClass().equals(YouTubeVideo.class)) {
			if (data.size() == 1) {
				YouTubeVideo video = (YouTubeVideo) data.get(0);
				try {
					if (video.isPremiere()) {
						this.type = NotificationType.YoutubePremiere;
						this.title = URLEncoder.encode(
								video.getChannelName() + " has scheduled a premiere for " + video.getPremiereDate(),
								StandardCharsets.UTF_8.toString());
					} else if (video.isShorts()) {
						this.type = NotificationType.YoutubeShort;
						this.title = URLEncoder.encode(video.getChannelName() + " has posted a Short!",
								StandardCharsets.UTF_8.toString());
					} else {
						this.type = NotificationType.Youtube;
						this.title = URLEncoder.encode(video.getChannelName() + " has uploaded!",
								StandardCharsets.UTF_8.toString());
					}
					this.message = URLEncoder.encode(video.getName(), StandardCharsets.UTF_8.toString());
					this.url = URLEncoder.encode("https://youtube.com/watch?v=" + video.getId(),
							StandardCharsets.UTF_8.toString());
				} catch (UnsupportedEncodingException e) {
					Logging.logger.severe(e.toString());
					System.exit(1);
				}

			} else if (data.size() > 1) {
				this.title = "";
				for (int i = 0; i < data.size(); i++) {
					YouTubeVideo video = (YouTubeVideo) data.get(i);
					this.title = this.title + video.getChannelName() + ", ";
				}
				try {
					this.title = URLEncoder.encode(this.title.substring(0, this.title.length() - 2) + " have uploaded!",
							StandardCharsets.UTF_8.toString());
					this.message = URLEncoder.encode("Check out their latest YoutTube videos!",
							StandardCharsets.UTF_8.toString());
					this.url = URLEncoder.encode("https://youtube.com/feed/subscriptions/",
							StandardCharsets.UTF_8.toString());
				} catch (UnsupportedEncodingException e) {
					Logging.logger.severe(e.toString());
					System.exit(1);
				}
			}
		} else if (data.get(0).getClass().equals(YTLiveStream.class)) {
			this.type = NotificationType.YoutubeStream;
			if (data.size() == 1) {
				YTLiveStream ytLiveStream = (YTLiveStream) data.get(0);
				try {
					this.title = URLEncoder.encode(ytLiveStream.getChannelName() + " has gone live on YouTube!",
							StandardCharsets.UTF_8.toString());
					this.message = URLEncoder.encode(ytLiveStream.getStreamName(), StandardCharsets.UTF_8.toString());
					this.url = URLEncoder.encode(
							"https://www.youtube.com/channel/" + ytLiveStream.getChannelId() + "/live/",
							StandardCharsets.UTF_8.toString());
				} catch (UnsupportedEncodingException e) {

				}
			} else if (data.size() > 1) {
				for (int i = 0; i < data.size(); i++) {
					YTLiveStream channel = (YTLiveStream) data.get(i);
					this.title = this.title + channel.getChannelName() + ", ";
				}
				try {
					this.title = URLEncoder.encode(
							this.title.substring(0, this.title.length() - 2) + " are live on YouTube!",
							StandardCharsets.UTF_8.toString());
					this.message = URLEncoder.encode("Check them out on YouTube!", StandardCharsets.UTF_8.toString());
					this.url = URLEncoder.encode("https://youtube.com/feed/subscriptions/",
							StandardCharsets.UTF_8.toString());
				} catch (UnsupportedEncodingException e) {
					Logging.logger.severe(e.toString());
					System.exit(1);
				}
			}
		}
	}

	public Notification(TwitchLiveStream stream) {
		this.type = NotificationType.Twitch;
		String channelName = stream.getChannelName();
		try {
			this.title = URLEncoder.encode(channelName + " is live on Twitch!", StandardCharsets.UTF_8.toString());
			this.message = URLEncoder.encode(stream.getStreamName(), StandardCharsets.UTF_8.toString());
			this.url = URLEncoder.encode("https://twitch.tv/" + channelName, StandardCharsets.UTF_8.toString());
		} catch (UnsupportedEncodingException e) {
			Logging.logger.severe(e.toString());
			System.exit(1);
		}
	}

	@Override
	public String toString() {
		return "Notification [title=" + title + ", message=" + message + ", url=" + url + ", type=" + type + "]";
	}

	public boolean send(AlertzyAuth auth) {
		String keysStr = "";
		for (String key : auth.getAccountKeys()) {
			keysStr += key + "_";
		}
		keysStr = keysStr.substring(0, keysStr.length() - 1);
		String url = "https://alertzy.app/send?accountKey=" + keysStr + "&title=" + this.title + "&message="
				+ this.message + "&link=" + this.url;
		OkHttpClient client = new OkHttpClient();
		Request request = new Request.Builder().url(url).build();
		try (Response response = client.newCall(request).execute()) {
			// Parse the HTTP response and makes sure it was sent
			JSONParser parser = new JSONParser();
			JSONObject responseJSON = (JSONObject) parser.parse(response.body().string());
			String success = responseJSON.get("response").toString();
			if (success.equalsIgnoreCase("success")) {
				Logging.logger.info("Sent Alertzy Notification: " + this.toString());
				return true;
			} else {
				Logging.logger.severe("Error sending Alertzy Notification: Sent to: " + responseJSON.get("sentTo")
						+ ", Error: " + responseJSON.get("error"));
				return false;
			}
		} catch (IOException | ParseException e) {
			Logging.logger.severe(e.toString());
			return false;
		}
	}
}

File: src/main/kotlin/com/jstockley/bsn/notification/Notification.kt
package com.jstockley.bsn.notification

import com.jstockley.bsn.AlertzyException
import com.jstockley.bsn.AlertzyFailException
import com.jstockley.bsn.AlertzyMixedException
import com.jstockley.bsn.twitch.TwitchLiveStream
import com.jstockley.bsn.youtube.live.YouTubeLiveStream
import com.jstockley.bsn.youtube.video.YouTubeVideo
import mu.KotlinLogging
import org.json.JSONObject
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets

class Notification() {

    private val logger = KotlinLogging.logger{}

    private lateinit var title: String

    private lateinit var message: String

    private lateinit var url: String

    private lateinit var type: NotificationType

    private var priority: Int = 0

    fun send(accountKeys: List<String>): Boolean{
        try {
            val keys: String = accountKeys.joinToString(separator="_"){ it }
            val requestURL = "https://alertzy.app/send?accountKey=${keys}&title=${this.title}&message=${this.message}&priority=${this.priority}&link=${this.url}"
            val client = HttpClient.newBuilder().build()
            val request = HttpRequest.newBuilder()
                .uri(URI.create(requestURL))
                .build()
            val response = client.send(request, HttpResponse.BodyHandlers.ofString())
            val responseJson =  JSONObject(response.body())
            val message: String = responseJson.getString("response")
            if (message == "fail") {
                throw AlertzyFailException(responseJson.toString(), "jack@jstockley.com")
            } else if (message == "mixed") {
                throw AlertzyMixedException(responseJson.toString(), "jack@jstockley.com")
            }
            logger.info { "Sent Notification: $this" }
            return true
        } catch (e: Exception) {
            val msg = e.stackTraceToString()
            logger.error { "Error sending Alertzy com.jstockley.Notification: $msg" }
            throw AlertzyException(msg, "jack@jstockley.com")
        }
    }

    override fun toString(): String {
        return "Notification(title='$title', message='$message', url='$url', type=$type, priority=$priority)"
    }

    constructor(title: String, message: String, type: NotificationType) : this() {
        this.title = URLEncoder.encode(title, StandardCharsets.UTF_8)
        this.message = URLEncoder.encode(message, StandardCharsets.UTF_8)
        this.url = ""
        this.type = type
        when (type) {
            NotificationType.StartingMessage -> {
                this.priority = 1
            }
            NotificationType.Error -> {
                this.priority = 2
            }
            else -> {
                this.priority = 0
            }
        }
    }

    constructor(data: List<Any>): this() {
        if(data[0]::class.java == YouTubeVideo::class.java){
            if (data.size == 1) {
                val video: YouTubeVideo = data[0] as YouTubeVideo
                if (video.isShort()) {
                    this.title = URLEncoder.encode("${video.getChannelName()} has uploaded a short!", StandardCharsets.UTF_8)
                    this.type = NotificationType.YouTubeShort
                } else if (video.isPremiere()) {
                    this.title = URLEncoder.encode("${video.getChannelName()} has scheduled a premiere for ${video.getPremiereDate()}", StandardCharsets.UTF_8)
                    this.type = NotificationType.YouTubePremiere
                } else {
                    this.title = URLEncoder.encode("${video.getChannelName()} has uploaded a video!", StandardCharsets.UTF_8)
                    this.type = NotificationType.YouTubeVideo
                }
                this.message = URLEncoder.encode(video.getVideoName(), StandardCharsets.UTF_8)
                this.url = URLEncoder.encode("https://youtube.com/watch?v=${video.getVideoId()}", StandardCharsets.UTF_8)
            } else {
                var titleString = ""
                for (i in 1 until (data.size)) {
                    val video = data[i] as YouTubeVideo
                    titleString += ("${video.getChannelName()}, ")
                }
                this.title = URLEncoder.encode("${titleString.substring(0, titleString.length - 2)} have uploaded a video!", StandardCharsets.UTF_8)
                this.message = URLEncoder.encode("Check out their latest YouTube videos!", StandardCharsets.UTF_8)
                this.url = URLEncoder.encode("https://youtube.com/feed/subscriptions/", StandardCharsets.UTF_8)
                this.type = NotificationType.YouTubeVideos
            }
        } else if (data[0]::class.java == YouTubeLiveStream::class.java) {
            this.type = NotificationType.YouTubeLiveStream
            if (data.size == 1) {
                val stream: YouTubeLiveStream = data[0] as YouTubeLiveStream
                this.title = URLEncoder.encode("${stream.getChannel()} has gone live on YouTube!", StandardCharsets.UTF_8)
                this.message = URLEncoder.encode(stream.getTitle(), StandardCharsets.UTF_8)
                this.url = URLEncoder.encode("https://youtube.com/channel/${stream.getChannelId()}/live/", StandardCharsets.UTF_8)
            } else {
                var titleString = ""
                for (i in 1..data.size) {
                    val stream = data[i] as YouTubeLiveStream
                    titleString += ("${stream.getChannel()}, ")
                }
                this.title = URLEncoder.encode("${titleString.substring(0, titleString.length - 2)} have gone live on YouTube!", StandardCharsets.UTF_8)
                this.message = URLEncoder.encode("Check out their YouTube live streams!", StandardCharsets.UTF_8)
                this.url = URLEncoder.encode("https://youtube.com/feed/subscriptions/", StandardCharsets.UTF_8)
            }
        }
    }

    constructor(data: TwitchLiveStream): this() {
        this.title = URLEncoder.encode("${data.getName()} has gone live on Twitch!", StandardCharsets.UTF_8)
        this.message = URLEncoder.encode("${data.getTitle()}: ${data.getCategory()}", StandardCharsets.UTF_8)
        this.url = URLEncoder.encode("https://twitch.tv/${data.getName()}", StandardCharsets.UTF_8)
        this.type = NotificationType.Twitch
    }
}